### PR TITLE
EE-1217: add get_call_stack, check immediate caller in auction

### PR DESCRIPTION
**Pull request by henrytill**
This PR introduces a `get_call_stack` function to the `runtime` module of the Rust contract API library.  The supporting host-side logic is also used to check/validate the immediate caller of the public API methods in the auction.

### CODE REVIEW

File: smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                    contract_package_hash.serialized_length()
+                }
+            }
+    }
+}
+
+impl FromBytes for ContractAddress {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (tag, remainder): (u8, &[u8]) = FromBytes::from_bytes(bytes)?;
+        match tag {
```
Comment (mpapierski) [inline smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/lib.rs:75]:
```suggestion
        let tag = ContractAddressTag::from_u8(tag).ok_or(Error::Formatting)?;
        match tag {
```

See also #1394

File: types/src/system/mod.rs [hunk 2]
```diff
@@ -194,3 +204,176 @@ mod system_contract_type {
...
+                    contract_hash,
+                } => contract_package_hash.serialized_length() + contract_hash.serialized_length(),
+            }
+    }
+}
+
+impl FromBytes for CallStackElement {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (tag, remainder): (u8, &[u8]) = FromBytes::from_bytes(bytes)?;
+        match tag {
```
Comment (mpapierski) [inline types/src/system/mod.rs:339]: See also #1394 

```suggestion
        let tag = CallStackElementTag::from_u8(tag).ok_or(bytesrepr::Error::Formatting)?;
        match tag {
```

File: types/src/system/auction/mod.rs [hunk 3]
```diff
@@ -98,9 +98,9 @@ pub trait Auction:
         amount: U512,
     ) -> Result<U512, Error> {
         let account_hash = AccountHash::from_public_key(&public_key, |x| self.blake2b(x));
-        if self.get_caller() != account_hash {
-            return Err(Error::InvalidPublicKey);
-        }
+        if self.get_immediate_caller() != Some(&CallStackElement::Session { account_hash }) {
```
Comment (goral09) [inline types/src/system/auction/mod.rs:101]: If in my deploy, I call a contract that calls an auction endpoint on my behalf – will that call fail now?

File: execution_engine/src/core/runtime/auction_internal.rs [hunk 4]
```diff
@@ -148,6 +151,13 @@ where
         self.context.get_caller()
     }
 
+    fn get_immediate_caller(&self) -> Option<&CallStackElement> {
+        let call_stack = self.call_stack();
+        let mut call_stack_iter = call_stack.iter().rev();
+        call_stack_iter.next();
+        call_stack_iter.next()
```
Comment (goral09) [inline execution_engine/src/core/runtime/auction_internal.rs:158]: Can you add a comment why you need to call `call_stack_iter.next()` twice before returning?

File: execution_engine/src/core/engine_state/executable_deploy_item.rs [hunk 5]
```diff
@@ -678,34 +744,68 @@ impl Distribution<ExecutableDeployItem> for Standard {
...
+            self.kind,
+            self.entry_point.entry_point_type(),
+            self.is_stored,
+        ) {
+            (DeployKind::Session, EntryPointType::Contract, _) => {
+                Err(Error::InvalidDeployItemVariant(
+                    "Contract deploy item has invalid 'Session' kind".to_string(),
+                ))
+            }
+            (DeployKind::Session, EntryPointType::Session, false) => {
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine/src/core/engine_state/executable_deploy_item.rs:782]: @dwerner and I had some hesitation about adding another set of identifiers into the mix, so we opted to leave as-is.

File: execution_engine/src/core/engine_state/mod.rs [hunk 6]
```diff
@@ -24,6 +24,7 @@ use std::{
     rc::Rc,
 };
 
+use executable_deploy_item::DeployKind;
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine/src/core/engine_state/mod.rs]: Addressed

File: execution_engine/src/core/runtime/auction_internal.rs [hunk 7]
```diff
@@ -148,6 +151,13 @@ where
         self.context.get_caller()
     }
 
+    fn get_immediate_caller(&self) -> Option<&CallStackElement> {
+        let call_stack = self.call_stack();
+        let mut call_stack_iter = call_stack.iter().rev();
+        call_stack_iter.next();
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine/src/core/runtime/auction_internal.rs]: Addressed

File: execution_engine_testing/tests/src/test/contract_api/get_call_stack.rs [hunk 8]
```diff
@@ -0,0 +1,3952 @@
+use get_call_stack_recursive_subcall::{Call, ContractAddress};
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine_testing/tests/src/test/contract_api/get_call_stack.rs]: Addressed

File: smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+const CONTRACT_PACKAGE_NAME: &str = "forwarder";
+const PACKAGE_ACCESS_KEY_NAME: &str = "forwarder_access";
+
+const CONTRACT_NAME: &str = "our_contract_name";
+
+const METHOD_FORWARDER_CONTRACT_NAME: &str = "forwarder_contract";
+const METHOD_FORWARDER_SESSION_NAME: &str = "forwarder_session";
+
+const ARG_CALLS: &str = "calls";
```
[the remark this answers is not in the document]
Comment (henrytill) [reply smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/main.rs]: Addressed

File: smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+use casper_types::{
+    bytesrepr,
+    bytesrepr::{Error, FromBytes, ToBytes, U8_SERIALIZED_LENGTH},
+    runtime_args, ApiError, CLType, CLTyped, ContractHash, ContractPackageHash, EntryPointType,
+    Key, Phase, RuntimeArgs, Tagged, URef, U512,
+};
+
+const DEFAULT_PAYMENT: u64 = 1_500_000_000_000;
+
+const ARG_CALLS: &str = "calls";
```
[the remark this answers is not in the document]
Comment (henrytill) [reply smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/lib.rs]: Addressed

File: execution_engine_testing/tests/src/test/contract_api/get_call_stack.rs [hunk 11]
```diff
@@ -0,0 +1,3952 @@
...
+const CONTRACT_RECURSIVE_SUBCALL: &str = "get_call_stack_recursive_subcall.wasm";
+const CONTRACT_CALL_RECURSIVE_SUBCALL: &str = "get_call_stack_call_recursive_subcall.wasm";
+
+const CONTRACT_PACKAGE_NAME: &str = "forwarder";
+const CONTRACT_NAME: &str = "our_contract_name";
+
+const CONTRACT_FORWARDER_ENTRYPOINT_CONTRACT: &str = "forwarder_contract";
+const CONTRACT_FORWARDER_ENTRYPOINT_SESSION: &str = "forwarder_session";
+
+const ARG_CALLS: &str = "calls";
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine_testing/tests/src/test/contract_api/get_call_stack.rs]: Addressed

File: execution_engine_testing/tests/src/test/regression/mod.rs [hunk 12]
```diff
@@ -24,7 +25,6 @@ mod ee_550;
 mod ee_572;
 mod ee_584;
 mod ee_597;
-mod ee_598;
```
[the remark this answers is not in the document]
Comment (henrytill) [reply execution_engine_testing/tests/src/test/regression/mod.rs]: Re-included and passes just fine.  Likely disabled mid-development when something came out of alignment, but everything is fine now.

File: smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/lib.rs (hunk 1 above)
Comment (henrytill) [reply smart_contracts/contracts/test/get-call-stack-recursive-subcall/src/lib.rs:82]: Addressed

File: types/src/system/mod.rs (hunk 2 above)
Comment (henrytill) [reply types/src/system/mod.rs:344]: Addressed

File: smart_contracts/contract/src/ext_ffi.rs [hunk 13]
```diff
@@ -700,6 +700,11 @@ extern "C" {
         out_ptr: *mut u8,
         out_size: usize,
     ) -> i32;
+    ///
```
Comment (mpapierski) [inline smart_contracts/contract/src/ext_ffi.rs:703]: This one stands out without rustdoc

**Review by mpapierski — APPROVED** [review]
Looks good to me 👍

File: types/src/system/auction/mod.rs (hunk 3 above)
Comment (henrytill) [reply types/src/system/auction/mod.rs:101]: > If, in my deploy's session code, I call a contract that calls an auction endpoint on my behalf – will that call fail now? 

Yes.

> Alternatively, is it only possible to call auction contract directly now? 

Yes.

> Will things like `auction_manager` (that cspr.live is using for delegate/undelegate) contract stop working?

Yes.

Further discussion on the subject at standup.

File: smart_contracts/contract/src/ext_ffi.rs (hunk 13 above)
Comment (henrytill) [reply smart_contracts/contract/src/ext_ffi.rs:703]: I'm following the example of the other `load_` function. :stuck_out_tongue_winking_eye:

File: execution_engine_testing/tests/src/test/regression/ee_1217_mint_proxy.rs [hunk 14]
```diff
@@ -88,47 +141,17 @@ fn faucet_should_create_account() {
...
-            .expect("should have last exec result");
-        let exec_response = response.last().expect("should have response");
-        exec_response.as_error().expect("should have error")
-    };
-    assert_matches!(
-        error,
-        engine_state::Error::Exec(execution::Error::Revert(ApiError::User(1)))
-    );
+    match builder.get_error() {
+        Some(engine_state::Error::Exec(execution::Error::Revert(ApiError::Mint(20)))) => {}
```
Comment (mpapierski) [inline execution_engine_testing/tests/src/test/regression/ee_1217_mint_proxy.rs:147]: If possible; for better readability

```suggestion
        Some(engine_state::Error::Exec(execution::Error::Revert(api_error))) if api_error == ApiError::from(mint::Error::InvalidContext) => {}
```

**Review by mpapierski — APPROVED** [review]

File: execution_engine_testing/tests/src/test/regression/ee_1217_mint_proxy.rs (hunk 14 above)
Comment (henrytill) [reply execution_engine_testing/tests/src/test/regression/ee_1217_mint_proxy.rs:147]: :+1:

Comment (henrytill) [reply execution_engine_testing/tests/src/test/regression/ee_1217_mint_proxy.rs]: Addressed

**Review by EdHastingsCasperLabs — APPROVED** [review]

**Comment by henrytill** [comment]
bors try

**Pull request merged**